In [22]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

`pandas` digunakan untuk membaca dan mengolah dataset CSV.

`numpy` digunakan untuk operasi numerik.


`train_test_split` digunakan untuk membagi dataset menjadi data training dan testing.


`StandardScaler` digunakan untuk normalisasi/standardisasi data.


`confusion_matrix`, `classification_report`, dan `accuracy_score` digunakan untuk evaluasi model.

In [23]:
df = pd.read_csv("Social_Network_Ads.csv")

print("First 5 rows:")
display(df.head())

print("\nDataset shape:")
print(df.shape)

print("\nDataset information:")
print(df.info())

print("\nMissing values:")
print(df.isnull().sum())

print("\nClass distribution:")
print(df["Purchased"].value_counts())

First 5 rows:


,User ID,Gender,Age,EstimatedSalary,Purchased
0,15624510,Male,19,19000,0
1,15810944,Male,35,20000,0
2,15668575,Female,26,43000,0
3,15603246,Female,27,57000,0
4,15804002,Male,19,76000,0



Dataset shape:
(400, 5)

Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 15.8+ KB
None

Missing values:
User ID            0
Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

Class distribution:
Purchased
0    257
1    143
Name: count, dtype: int64


Membaca dataset menggunakan `pd.read_csv()` dan mengecek dataset menggunakan beberapa fungsi:

| Kode                             | Fungsi                                |
| -------------------------------- | ------------------------------------- |
| `df.head()`                      | Menampilkan 5 baris pertama           |
| `df.shape`                       | Melihat jumlah baris dan kolom        |
| `df.info()`                      | Melihat tipe data setiap kolom        |
| `df.isnull().sum()`              | Mengecek missing value                |
| `df["Purchased"].value_counts()` | Melihat jumlah data pada setiap kelas |


In [24]:
df["Gender"] = df["Gender"].map({
    "Male": 1,
    "Female": 0
})

Mengubah kolom Gender awalnya berbentuk teks, yaitu Male dan Female menjadi angka.

* Male: 1
* Female: 0


In [25]:
X = df[["Gender", "Age", "EstimatedSalary"]]
y = df["Purchased"]

print("Features:")
display(X.head())

print("\nTarget:")
display(y.head())

Features:


,Gender,Age,EstimatedSalary
0,1,19,19000
1,1,35,20000
2,0,26,43000
3,0,27,57000
4,1,19,76000



Target:


,Purchased
0,0
1,0
2,0
3,0
4,0


Memisahkan dataset menjadi fitur input dan target output.

In [27]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("X_test shape :", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape :", y_test.shape)

X_train shape: (320, 3)
X_test shape : (80, 3)
y_train shape: (320,)
y_test shape : (80,)


Dataset dibagi menjadi data training dan data testing.

80% data → training

20% data → testing


Data training digunakan untuk melatih model ANN.
Data testing digunakan untuk menguji performa model terhadap data baru.

Parameter `stratify=y` digunakan agar proporsi kelas Purchased = 0 dan Purchased = 1 tetap seimbang pada data training dan testing.

In [28]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

X_train_scaled = X_train_scaled.T
X_test_scaled = X_test_scaled.T

y_train = y_train.values.reshape(1, -1)
y_test = y_test.values.reshape(1, -1)

print("X_train_scaled shape:", X_train_scaled.shape)
print("X_test_scaled shape :", X_test_scaled.shape)
print("y_train shape:", y_train.shape)
print("y_test shape :", y_test.shape)

X_train_scaled shape: (3, 320)
X_test_scaled shape : (3, 80)
y_train shape: (1, 320)
y_test shape : (1, 80)


Digunakan untuk melakukan standardisasi data menggunakan StandardScaler.

Standardisasi diperlukan karena fitur memiliki skala yang berbeda. Contohnya:

* Gender hanya bernilai 0 atau 1
* Age bernilai puluhan
* EstimatedSalary bernilai puluhan ribu sampai ratusan ribu

`fit_transform()` digunakan pada data training agar scaler menghitung nilai rata-rata dan standar deviasi dari data training, lalu mengubah data training tersebut.

`transform()` digunakan pada data testing agar data testing menggunakan skala yang sama dengan data training.

In [29]:
def sigmoid(Z):
    return 1 / (1 + np.exp(-Z))

def sigmoid_derivative(A):
    return A * (1 - A)

Digunakan untuk mendefinisikan fungsi aktivasi sigmoid dan turunannya.

Fungsi sigmoid digunakan untuk mengubah nilai input menjadi rentang 0 sampai 1.

Rumus sigmoid:

`sigmoid(x) = 1 / (1 + e^(-x))`

Karena output berada pada rentang 0 sampai 1, sigmoid cocok digunakan untuk binary classification.

Fungsi `sigmoid_derivative()` digunakan pada proses backpropagation untuk menghitung gradient pada hidden layer.

In [30]:
def initialize_parameters(input_size, hidden_size, output_size):
    np.random.seed(42)

    W1 = np.random.randn(hidden_size, input_size) * 0.01
    b1 = np.zeros((hidden_size, 1))

    W2 = np.random.randn(output_size, hidden_size) * 0.01
    b2 = np.zeros((output_size, 1))

    parameters = {
        "W1": W1,
        "b1": b1,
        "W2": W2,
        "b2": b2
    }

    return parameters

Digunakan untuk menginisialisasi parameter awal model ANN, yaitu bobot dan bias.

Model yang digunakan memiliki struktur:

3 input → 4 hidden neuron → 1 output neuron

Karena terdapat 3 input dan 4 neuron pada hidden layer, maka bobot W1 memiliki ukuran:

4 × 3

Bias b1 memiliki ukuran:

4 × 1

Karena output layer memiliki 1 neuron dan menerima input dari 4 hidden neuron, maka bobot W2 memiliki ukuran:

1 × 4

Bias b2 memiliki ukuran:

1 × 1

Bobot diberi nilai awal random kecil, sedangkan bias diberi nilai 0. Nilai awal ini kemudian diperbarui selama proses training.

In [31]:
def forward_propagation(X, parameters):
    W1 = parameters["W1"]
    b1 = parameters["b1"]
    W2 = parameters["W2"]
    b2 = parameters["b2"]

    Z1 = np.dot(W1, X) + b1
    A1 = sigmoid(Z1)

    Z2 = np.dot(W2, A1) + b2
    A2 = sigmoid(Z2)

    cache = {
        "Z1": Z1,
        "A1": A1,
        "Z2": Z2,
        "A2": A2
    }

    return A2, cache

Bagian ini berisi proses forward propagation, yaitu proses perhitungan dari input menuju output.

Tahap pertama:

Z1 = W1 × X + b1

A1 = sigmoid(Z1)

Nilai input dikalikan dengan bobot pertama, ditambahkan bias, lalu diproses menggunakan fungsi aktivasi sigmoid. Hasilnya adalah output hidden layer.

Tahap kedua:

Z2 = W2 × A1 + b2

A2 = sigmoid(Z2)

Output hidden layer kemudian dikalikan dengan bobot kedua, ditambahkan bias output, lalu diproses menggunakan sigmoid.

A2 adalah hasil prediksi probabilitas model.

Nilai cache disimpan karena akan digunakan kembali pada proses `backpropagation`.

In [32]:
def compute_cost(Y, A2):
    m = Y.shape[1]
    epsilon = 1e-8

    cost = -np.sum(
        Y * np.log(A2 + epsilon) + (1 - Y) * np.log(1 - A2 + epsilon)
    ) / m

    return np.squeeze(cost)

Digunakan untuk menghitung nilai error atau loss model menggunakan binary cross-entropy.

Binary cross-entropy cocok digunakan karena target klasifikasi hanya terdiri dari dua kelas, yaitu 0 dan 1.

Semakin kecil nilai cost, semakin baik model dalam melakukan prediksi.

Variabel epsilon digunakan untuk mencegah error matematika ketika terjadi log(0).

Cost dihitung berdasarkan perbedaan antara label asli Y dan hasil prediksi probabilitas A2.

In [33]:
def backward_propagation(X, Y, parameters, cache):
    m = X.shape[1]

    W2 = parameters["W2"]
    A1 = cache["A1"]
    A2 = cache["A2"]

    dZ2 = A2 - Y
    dW2 = np.dot(dZ2, A1.T) / m
    db2 = np.sum(dZ2, axis=1, keepdims=True) / m

    dZ1 = np.dot(W2.T, dZ2) * sigmoid_derivative(A1)
    dW1 = np.dot(dZ1, X.T) / m
    db1 = np.sum(dZ1, axis=1, keepdims=True) / m

    grads = {
        "dW1": dW1,
        "db1": db1,
        "dW2": dW2,
        "db2": db2
    }

    return grads

Bagian ini berisi proses backpropagation, yaitu proses menghitung gradient untuk memperbaiki bobot dan bias.

Pertama, error pada output layer dihitung:

`dZ2 = A2 - Y`

Kemudian gradient untuk bobot dan bias output dihitung:

`dW2`

`db2`

Setelah itu, error dihitung mundur ke hidden layer:

`dZ1`

Kemudian dihitung gradient untuk bobot dan bias hidden layer:

`dW1`

`db1`

Nilai gradient ini akan digunakan untuk memperbarui parameter model pada tahap berikutnya.

In [34]:
def update_parameters(parameters, grads, learning_rate):
    parameters["W1"] = parameters["W1"] - learning_rate * grads["dW1"]
    parameters["b1"] = parameters["b1"] - learning_rate * grads["db1"]

    parameters["W2"] = parameters["W2"] - learning_rate * grads["dW2"]
    parameters["b2"] = parameters["b2"] - learning_rate * grads["db2"]

    return parameters

Bagian ini digunakan untuk memperbarui bobot dan bias menggunakan metode gradient descent.

Rumus dasarnya adalah:

`parameter baru = parameter lama - learning_rate × gradient`

learning_rate menentukan seberapa besar langkah pembaruan parameter.

Jika learning rate terlalu kecil, proses belajar bisa lambat. Jika terlalu besar, model bisa tidak stabil.

Pada bagian ini, parameter `W1`, `b1`, `W2`, dan `b2` diperbarui berdasarkan gradient yang dihitung pada proses backpropagation.

In [35]:
def train_neural_network(X, Y, input_size, hidden_size, output_size, epochs, learning_rate):
    parameters = initialize_parameters(input_size, hidden_size, output_size)

    cost_history = []

    for i in range(epochs):
        A2, cache = forward_propagation(X, parameters)
        cost = compute_cost(Y, A2)
        grads = backward_propagation(X, Y, parameters, cache)
        parameters = update_parameters(parameters, grads, learning_rate)

        cost_history.append(cost)

        if i % 1000 == 0:
            print(f"Epoch {i}: Cost = {cost:.6f}")

    return parameters, cost_history

Bagian ini menggabungkan seluruh proses training ANN.

Dalam setiap epoch, model melakukan:

1. Forward propagation
2. Menghitung cost
3. Backpropagation
4. Update parameter

Proses ini dilakukan berulang-ulang sampai jumlah epoch selesai.

Nilai cost disimpan ke dalam cost_history agar perubahan error selama training dapat diamati.

Setiap 1000 epoch, program menampilkan nilai cost. Jika cost menurun, berarti model sedang belajar dari data.

In [36]:
trained_parameters, cost_history = train_neural_network(
    X_train_scaled,
    y_train,
    input_size=3,
    hidden_size=4,
    output_size=1,
    epochs=10000,
    learning_rate=0.1
)

Epoch 0: Cost = 0.690345
Epoch 1000: Cost = 0.331883
Epoch 2000: Cost = 0.314666
Epoch 3000: Cost = 0.313070
Epoch 4000: Cost = 0.312513
Epoch 5000: Cost = 0.312226
Epoch 6000: Cost = 0.312065
Epoch 7000: Cost = 0.311969
Epoch 8000: Cost = 0.311909
Epoch 9000: Cost = 0.311868


Cell ini menjalankan proses training model ANN.

Parameter yang digunakan:

| Parameter     | Nilai | Keterangan                      |
| ------------- | ----: | ------------------------------- |
| input_size    |     3 | Gender, Age, EstimatedSalary    |
| hidden_size   |     4 | Jumlah neuron pada hidden layer |
| output_size   |     1 | Output binary classification    |
| epochs        | 10000 | Jumlah iterasi training         |
| learning_rate |   0.1 | Kecepatan pembelajaran model    |


Setelah training selesai, hasil parameter yang sudah dilatih disimpan dalam variabel:

`trained_parameters`

Parameter inilah yang digunakan untuk melakukan prediksi.

In [37]:
def predict(X, parameters):
    A2, _ = forward_propagation(X, parameters)
    predictions = (A2 >= 0.5).astype(int)
    return predictions, A2

Cell ini digunakan untuk membuat fungsi prediksi.

Model menghasilkan nilai probabilitas A2 dari fungsi sigmoid.

Aturan klasifikasi:

`A2 >= 0.5 → class 1 atau membeli`

`A2 < 0.5  → class 0 atau tidak membeli`

Fungsi ini mengembalikan dua nilai:

`predictions → hasil class 0 atau 1`

`A2 → nilai probabilitas prediksi`

In [38]:
y_pred_train, train_prob = predict(X_train_scaled, trained_parameters)
y_pred_test, test_prob = predict(X_test_scaled, trained_parameters)

train_accuracy = accuracy_score(y_train.flatten(), y_pred_train.flatten())
test_accuracy = accuracy_score(y_test.flatten(), y_pred_test.flatten())

print("Train Accuracy:", train_accuracy)
print("Test Accuracy :", test_accuracy)

Train Accuracy: 0.846875
Test Accuracy : 0.8125


Cell ini digunakan untuk mengevaluasi performa model pada data training dan data testing.

Hasil yang diperoleh:

`Train Accuracy: 0.846875`

`Test Accuracy : 0.8125`

Artinya, model memperoleh akurasi training sebesar:

`84,69%`

dan akurasi testing sebesar:

`81,25%`

Model cukup baik karena akurasi testing masih di atas 80%.

Perbedaan antara train accuracy dan test accuracy juga tidak terlalu besar, sehingga model tidak menunjukkan overfitting yang parah.

In [39]:
cm = confusion_matrix(y_test.flatten(), y_pred_test.flatten())

print("Confusion Matrix:")
print(cm)

print("\nClassification Report:")
print(classification_report(y_test.flatten(), y_pred_test.flatten()))

Confusion Matrix:
[[43  8]
 [ 7 22]]

Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.84      0.85        51
           1       0.73      0.76      0.75        29

    accuracy                           0.81        80
   macro avg       0.80      0.80      0.80        80
weighted avg       0.81      0.81      0.81        80



Cell ini digunakan untuk mengevaluasi hasil klasifikasi secara lebih detail.

Hasil confusion matrix:

[[43  8]

 [ 7 22]]

 | Aktual / Prediksi | Prediksi 0 | Prediksi 1 |
| ----------------- | ---------: | ---------: |
| Aktual 0          |         43 |          8 |
| Aktual 1          |          7 |         22 |

Penjelasan:

| Jenis          | Jumlah | Arti                                            |
| -------------- | -----: | ----------------------------------------------- |
| True Negative  |     43 | User tidak membeli dan diprediksi tidak membeli |
| False Positive |      8 | User tidak membeli tetapi diprediksi membeli    |
| False Negative |      7 | User membeli tetapi diprediksi tidak membeli    |
| True Positive  |     22 | User membeli dan diprediksi membeli             |


Total data testing:

`80 data`

Prediksi benar:

`43 + 22 = 65 data`

Prediksi salah:

`8 + 7 = 15 data`

Akurasi:

`65 / 80 = 0,8125 atau 81,25%`

Classification report menunjukkan bahwa class 0 memiliki performa lebih baik dibanding class 1. Hal ini terlihat dari nilai precision, recall, dan f1-score class 0 yang lebih tinggi.

In [40]:
test_result = X_test.copy()
test_result["Actual"] = y_test.flatten()
test_result["Probability"] = test_prob.flatten()
test_result["Prediction"] = y_pred_test.flatten()
test_result["Correct"] = (test_result["Actual"] == test_result["Prediction"]).astype(int)

display(test_result.head(20))

,Gender,Age,EstimatedSalary,Actual,Probability,Prediction,Correct
331,0,48,119000,1,0.887707,1,1
92,1,26,15000,0,0.000757,0,1
1,1,35,20000,0,0.002886,0,1
234,0,38,112000,0,0.757616,1,0
136,0,20,82000,0,0.001167,0,1
356,1,54,70000,1,0.881416,1,1
122,0,40,75000,0,0.485122,0,1
223,1,60,102000,1,0.898017,1,1
64,0,59,83000,0,0.894993,1,0
398,1,36,33000,0,0.008495,0,1


Cell ini digunakan untuk menampilkan hasil prediksi dalam bentuk tabel.

In [41]:
new_data = pd.DataFrame({
    "Gender": [1],              # Male = 1, Female = 0
    "Age": [35],
    "EstimatedSalary": [75000]
})

new_data_scaled = scaler.transform(new_data)
new_data_scaled = new_data_scaled.T

new_prediction, new_probability = predict(new_data_scaled, trained_parameters)

print("New Data:")
display(new_data)

print("Prediction Probability:", new_probability[0][0])
print("Prediction Class:", new_prediction[0][0])

if new_prediction[0][0] == 1:
    print("Result: User diprediksi membeli produk")
else:
    print("Result: User diprediksi tidak membeli produk")

New Data:


,Gender,Age,EstimatedSalary
0,1,35,75000


Prediction Probability: 0.14097853948895767
Prediction Class: 0
Result: User diprediksi tidak membeli produk


Cell ini digunakan untuk melakukan prediksi terhadap data baru.

Berdasarkan hasil pengujian, model Artificial Neural Network yang dibangun dari awal menggunakan NumPy menghasilkan train accuracy sebesar 84,69% dan test accuracy sebesar 81,25%. Hasil ini menunjukkan bahwa model mampu melakukan klasifikasi keputusan pembelian produk dengan cukup baik berdasarkan fitur Gender, Age, dan EstimatedSalary.

Confusion matrix menunjukkan bahwa dari 80 data testing, terdapat 43 data user yang tidak membeli dan berhasil diprediksi sebagai tidak membeli, serta 22 data user yang membeli dan berhasil diprediksi sebagai membeli. Model juga menghasilkan 8 false positive, yaitu user yang sebenarnya tidak membeli tetapi diprediksi membeli, dan 7 false negative, yaitu user yang sebenarnya membeli tetapi diprediksi tidak membeli.

Berdasarkan classification report, class 0 memiliki precision sebesar 0,86, recall sebesar 0,84, dan f1-score sebesar 0,85. Sementara itu, class 1 memiliki precision sebesar 0,73, recall sebesar 0,76, dan f1-score sebesar 0,75. Hal ini menunjukkan bahwa model lebih baik dalam mengklasifikasikan user yang tidak membeli dibandingkan user yang membeli.

Secara keseluruhan, model ANN from scratch ini memiliki performa yang cukup baik dengan akurasi testing sebesar 81,25%. Hasil ini juga menunjukkan bahwa proses ANN seperti forward propagation, perhitungan cost, backpropagation, dan update parameter dapat dilakukan secara manual menggunakan NumPy tanpa menggunakan library deep learning seperti TensorFlow atau Keras.